src/Playwright/Core/Waiter.cs

```diff
@@ -117,6 +117,18 @@ internal void RejectImmediately(Exception exception)
         _immediateError = exception;
     }
 
+    internal void RejectOnCancellation(CancellationToken cancellationToken)
+    {
+        if (!cancellationToken.CanBeCanceled)
+        {
+            return;
+        }
+
+        var cts = new TaskCompletionSource<bool>();
+        var registration = cancellationToken.Register(cts.SetCanceled, false);
```

Comment:
If you let the cancellation throw instead (https://github.com/microsoft/playwright-dotnet/pull/2597#discussion_r1211064250) this should use the `Set Canceled(cancellationToken)` instead so users can inspect it if the catch it higher up in the stack.